File: expiration-checker.py
#!/usr/bin/env python3
"""
Alma Expiration Checker (Python)

Finds users who expired before a user-supplied date and exports results to CSV/JSON.

Behavior:
- Prompts for Production or Sandbox environment and loads `.env` or `.env.sandbox`.
- Prompts for a cutoff date (YYYY-MM-DD) and finds users whose expiration date is strictly before that date.
- Fetches users via Alma Users API with pagination.
- Exports results to a timestamped CSV and JSON file.

If you haven't already, you'll need to use pip to install requests and python-dotenv. 

Save as `expiration-checker.py` and run with:

    pip install requests python-dotenv
    python3 expiration-checker.py

"""


# === Import required libraries ===
import os
import sys
import csv
import json
import time
from datetime import datetime, date
import requests
from requests.exceptions import RequestException
import re
from dotenv import load_dotenv

# === Configuration and network settings ===
# Timeout for HTTP requests (in seconds)
REQUEST_TIMEOUT = 30
# Number of users to fetch per API call (pagination)
PAGE_LIMIT = 50
# Limit number of users processed for quick testing (set to None to disable)
QUICK_SCAN_LIMIT = 1000
# Print the first user returned for debugging if True
DEBUG_SAMPLE = False

############################################################
# Function: select_environment
# Prompts the user to select Production or Sandbox environment.
# Loads the appropriate .env file for API credentials.
############################################################
def select_environment():
    print("\nSelect environment:")
    print("1. Production")
    print("2. Sandbox")
    while True:
        choice = input("\nEnter choice (1 or 2): ").strip()
        if choice == '1':
            if os.path.exists('.env'):
                load_dotenv('.env')
                print('Loaded Production environment')
                return 'production'
            else:
                print('Error: .env file not found')
                sys.exit(1)
        elif choice == '2':
            if os.path.exists('.env.sandbox'):
                load_dotenv('.env.sandbox')
                print('Loaded Sandbox environment')
                return 'sandbox'
            else:
                print('Error: .env.sandbox file not found')
                sys.exit(1)
        else:
            print('Invalid choice. Please enter 1 or 2.')

# This is the prompt for the cutoff date
############################################################
# Function: prompt_for_date
# Prompts the user to enter a cutoff date in YYYY-MM-DD format.
# Returns a datetime object representing the end of that day.
############################################################
def prompt_for_date():
    while True:
        s = input('\nEnter cutoff date (YYYY-MM-DD): ').strip()
        try:
            d = datetime.strptime(s, '%Y-%m-%d').date()
            return datetime(d.year, d.month, d.day, 23, 59, 59)
        except Exception:
            print('Invalid date format. Use YYYY-MM-DD.')


############################################################
# Function: parse_alma_date
# Tries to parse various Alma date formats into a datetime object.
# Returns None if parsing fails.
############################################################
def parse_alma_date(s):
    if not s:
        return None
    s = str(s).strip()
    # Try several common formats, return a datetime
    fmts = ['%Y-%m-%d', '%Y-%m-%dT%H:%M:%S.%fZ', '%Y-%m-%dT%H:%M:%SZ', '%Y-%m-%dT%H:%M:%S']
    for fmt in fmts:
        try:
            dt = datetime.strptime(s, fmt)
            # If format was date-only, dt will be at midnight; that's fine
            return dt
        except Exception:
            continue
    # Handle ISO with Z by replacing Z with +00:00 for fromisoformat
    try:
        if s.endswith('Z'):
            s2 = s.replace('Z', '+00:00')
            return datetime.fromisoformat(s2)
        return datetime.fromisoformat(s)
    except Exception:
        return None


############################################################
# Function: extract_email_from_contact
# Extracts the preferred email address from a user's contact info.
# Returns an empty string if not found.
############################################################
def extract_email_from_contact(contact_info):
    if not contact_info:
        return ''
    emails = contact_info.get('email', [])
    if not emails:
        return ''
    # prefer preferred email
    preferred = [e for e in emails if e.get('preferred')]
    if preferred:
        return preferred[0].get('email_address', '')
    return emails[0].get('email_address', '')


############################################################
# Function: get_all_users
# Generator that fetches all users from Alma via paginated API calls.
# Yields each user as a dictionary.
############################################################
def get_all_users(api_key, base_url, q=None):
    limit = PAGE_LIMIT
    offset = 0
    headers = {
        'Authorization': f'apikey {api_key}',
        'Accept': 'application/json'
    }

    while True:
        params = {
            'limit': limit,
            'offset': offset,
            'format': 'json',
            'expand': 'full'
        }
        if q:
            params['q'] = q
        url = f"{base_url.rstrip('/')}/almaws/v1/users"
        print(f"Fetching users: offset={offset} limit={limit}")
        try:
            resp = requests.get(url, headers=headers, params=params, verify=False, timeout=REQUEST_TIMEOUT)
        except RequestException as e:
            print(f"Request error: {e}")
            break
        if resp.status_code != 200:
            print(f"Failed to fetch users: {resp.status_code} {resp.text}")
            break
        data = resp.json()
        users = data.get('user', []) or []
        print(f"  Received {len(users)} users")
        if not users:
            break
        for user in users:
            yield user
        total = data.get('total_record_count')
        offset += limit
        if total is None:
            # no total provided - stop when fewer than limit returned
            if len(users) < limit:
                break
        else:
            if offset >= total:
                break
        # be kind to API
        time.sleep(0.2)


############################################################
# Function: sanitize_group_name
# Returns a filesystem-safe version of a group name for filenames.
############################################################
def sanitize_group_name(name: str) -> str:
    if not name:
        return 'group'
    s = name.strip()
    s = re.sub(r"\s+", "_", s)
    s = re.sub(r"[^A-Za-z0-9_\-]", "", s)
    return s or 'group'


############################################################
# Function: get_user_groups
# Fetches user groups from Alma configuration endpoint.
# Returns a list of group names.
############################################################
def get_user_groups(api_key, base_url):
    headers = {
        'Authorization': f'apikey {api_key}',
        'Accept': 'application/json'
    }
    groups = []
    offset = 0
    limit = 200
    while True:
        params = {'limit': limit, 'offset': offset}
        url = f"{base_url.rstrip('/')}/almaws/v1/conf/user_groups"
        try:
            resp = requests.get(url, headers=headers, params=params, verify=False, timeout=REQUEST_TIMEOUT)
        except RequestException as e:
            print(f"Failed to fetch user groups: {e}")
            break
        if resp.status_code != 200:
            print(f"Failed to fetch user groups: {resp.status_code} {resp.text}")
            break
        data = resp.json()
        items = data.get('user_group') or data.get('user_groups') or []
        if not items:
            break
        for g in items:
            name = g.get('value') or g.get('desc') or g.get('code') or g.get('name')
            if name:
                groups.append(str(name))
        total = data.get('total_record_count')
        offset += limit
        if total is None:
            if len(items) < limit:
                break
        else:
            if offset >= total:
                break
        time.sleep(0.2)
    return groups


############################################################
# Function: collect_expired_users
# Iterates through all users and collects those whose expiration date
# is before or equal to the cutoff date. Returns the list and total scanned.
############################################################
def collect_expired_users(api_key, base_url, cutoff_date, q=None, max_processed=None):
    results = []
    count = 0
    # Apply quick-scan limiter if provided and caller didn't pass a max
    if max_processed is None and QUICK_SCAN_LIMIT:
        max_processed = QUICK_SCAN_LIMIT
        print(f"Quick-scan limiter active: will stop after {max_processed} users")
    for user in get_all_users(api_key, base_url, q=q):
        count += 1
        if count == 1 and DEBUG_SAMPLE:
            try:
                print('\n--- Debug: sample user (first returned) ---')
                print('Available keys:', ', '.join(sorted(user.keys())))
                # Show commonly relevant fields
                sample = {
                    'primary_id': user.get('primary_id'),
                    'expiry_date': user.get('expiry_date'),
                    'expiration_date': user.get('expiration_date'),
                    'contact_info_keys': list(user.get('contact_info', {}).keys()) if isinstance(user.get('contact_info'), dict) else None,
                    'user_group': user.get('user_group')
                }
                print(json.dumps(sample, indent=2, default=str))
            except Exception as e:
                print('Debug sample error:', e)
        if max_processed and count > max_processed:
            print(f"Reached scanning limit ({max_processed}) - stopping early")
            break
        primary_id = user.get('primary_id') or user.get('primaryId') or user.get('primary')
        first_name = user.get('first_name') or user.get('firstName') or ''
        last_name = user.get('last_name') or user.get('lastName') or ''
        # expiration might be named several ways in different responses
        expiration_raw = user.get('expiration_date') or user.get('expiry_date') or user.get('expirationDate') or user.get('expiryDate') or user.get('expiration')
        expiration = parse_alma_date(expiration_raw)
        email = extract_email_from_contact(user.get('contact_info') or {})

        # Compare datetimes: include expiry on the cutoff date (<=)
        if expiration and cutoff_date and expiration <= cutoff_date:
            results.append({
                'primary_id': primary_id,
                'first_name': first_name,
                'last_name': last_name,
                'email': email,
                'expiration_date': expiration.isoformat()
            })

    return results, count


############################################################
# Function: save_csv
# Saves the list of expired users to a timestamped CSV file.
# Returns the filename.
############################################################
def save_csv(rows, environment, suffix=None):
    ts = datetime.now().strftime('%Y%m%d_%H%M%S')
    if suffix:
        filename = f"expired_users_{environment}_{suffix}_{ts}.csv"
    else:
        filename = f"expired_users_{environment}_{ts}.csv"
    fieldnames = ['primary_id', 'first_name', 'last_name', 'email', 'expiration_date']
    with open(filename, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        for r in rows:
            writer.writerow(r)
    return filename


############################################################
# Function: save_json
# Saves the list of expired users to a timestamped JSON file.
# Returns the filename.
############################################################
def save_json(rows, environment, suffix=None):
    ts = datetime.now().strftime('%Y%m%d_%H%M%S')
    if suffix:
        filename = f"expired_users_{environment}_{suffix}_{ts}.json"
    else:
        filename = f"expired_users_{environment}_{ts}.json"
    with open(filename, 'w', encoding='utf-8') as f:
        json.dump(rows, f, indent=2)
    return filename


############################################################
# Main program logic
# - Prints header
# - Loads environment and API credentials
# - Prompts for cutoff date
# - Scans users and collects expired ones
# - Saves results to CSV and JSON
############################################################
def main():
    print('=' * 60)
    print('ALMA EXPIRATION CHECKER (Python)')
    print('=' * 60)

    env = select_environment()
    api_key = os.getenv('ALMA_API_KEY')
    base_url = os.getenv('ALMA_API_BASE_URL')

    if not api_key or not base_url:
        print('Error: ALMA_API_KEY and ALMA_API_BASE_URL must be set in the chosen .env')
        sys.exit(1)

    # simple interactive flow: prompt for cutoff and scan all users
    cutoff = prompt_for_date()
    print(f"\nFinding users expired before {cutoff.isoformat()}...")

    rows, scanned = collect_expired_users(api_key, base_url, cutoff)
    print(f"\nScanned {scanned} users; found {len(rows)} expired users")

    if rows:
        csv_file = save_csv(rows, env)
        json_file = save_json(rows, env)
        print(f"\nSaved CSV: {csv_file}")
        print(f"Saved JSON: {json_file}")
    else:
        print('\nNo expired users found for that cutoff date.')

    print('\nDone!')


############################################################
# Script entry point
# - Suppresses SSL warnings for demo purposes
# - Ensures QUICK_SCAN_LIMIT is defined
# - Calls main()
############################################################
if __name__ == '__main__':
    import urllib3
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
    try:
        QUICK_SCAN_LIMIT
    except NameError:
        QUICK_SCAN_LIMIT = None
    main()

File: expiration-checker.go
package main

import (
	"bufio"
	"crypto/tls"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const (
	requestTimeout = 30 * time.Second
	pageLimit      = 50
	quickScanLimit = 1000 // set to 0 to disable
	debugSample    = false
)

type usersResponse struct {
	Users            []map[string]any `json:"user"`
	TotalRecordCount int              `json:"total_record_count"`
}

type expiredUserRow struct {
	PrimaryID      string `json:"primary_id"`
	FirstName      string `json:"first_name"`
	LastName       string `json:"last_name"`
	Email          string `json:"email"`
	PatronType     string `json:"patron_type"`
	ExpirationDate string `json:"expiration_date"`
}

func main() {
	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("ALMA EXPIRATION CHECKER (Go)")
	fmt.Println(strings.Repeat("=", 60))

	environment, err := selectEnvironment()
	if err != nil {
		fmt.Println("Error:", err)
		os.Exit(1)
	}

	apiKey := strings.TrimSpace(os.Getenv("ALMA_API_KEY"))
	baseURL := strings.TrimSpace(os.Getenv("ALMA_API_BASE_URL"))
	if apiKey == "" || baseURL == "" {
		fmt.Println("Error: ALMA_API_KEY and ALMA_API_BASE_URL must be set in the chosen .env")
		os.Exit(1)
	}

	cutoff := promptForDate()
	fmt.Printf("\nFinding users expired on or before %s...\n", cutoff.Format(time.RFC3339))

	rows, scanned, err := collectExpiredUsers(apiKey, baseURL, cutoff, quickScanLimit)
	if err != nil {
		fmt.Println("Error:", err)
		os.Exit(1)
	}

	fmt.Printf("\nScanned %d users; found %d expired users\n", scanned, len(rows))

	if len(rows) > 0 {
		csvFile, err := saveCSV(rows, environment)
		if err != nil {
			fmt.Println("Error saving CSV:", err)
			os.Exit(1)
		}

		jsonFile, err := saveJSON(rows, environment)
		if err != nil {
			fmt.Println("Error saving JSON:", err)
			os.Exit(1)
		}

		fmt.Printf("\nSaved CSV: %s\n", csvFile)
		fmt.Printf("Saved JSON: %s\n", jsonFile)
	} else {
		fmt.Println("\nNo expired users found for that cutoff date.")
	}

	fmt.Println("\nDone!")
}

func selectEnvironment() (string, error) {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("\nSelect environment:")
	fmt.Println("1. Production")
	fmt.Println("2. Sandbox")

	for {
		fmt.Print("\nEnter choice (1 or 2): ")
		if !scanner.Scan() {
			return "", fmt.Errorf("failed to read input")
		}
		choice := strings.TrimSpace(scanner.Text())

		switch choice {
		case "1":
			path, err := findEnvFile(".env")
			if err != nil {
				return "", err
			}
			if err := loadDotEnv(path); err != nil {
				return "", err
			}
			fmt.Println("Loaded Production environment")
			return "production", nil
		case "2":
			path, err := findEnvFile(".env.sandbox")
			if err != nil {
				return "", err
			}
			if err := loadDotEnv(path); err != nil {
				return "", err
			}
			fmt.Println("Loaded Sandbox environment")
			return "sandbox", nil
		default:
			fmt.Println("Invalid choice. Please enter 1 or 2.")
		}
	}
}

func promptForDate() time.Time {
	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print("\nEnter cutoff date (YYYY-MM-DD) for users expired on or before that date: ")
		if !scanner.Scan() {
			fmt.Println("Input cancelled.")
			os.Exit(1)
		}
		s := strings.TrimSpace(scanner.Text())
		d, err := time.Parse("2006-01-02", s)
		if err != nil {
			fmt.Println("Invalid date format. Use YYYY-MM-DD.")
			continue
		}
		return time.Date(d.Year(), d.Month(), d.Day(), 23, 59, 59, 0, time.Local)
	}
}

func findEnvFile(name string) (string, error) {
	searchDirs := []string{".", "..", "../..", "../../.."}
	for _, dir := range searchDirs {
		candidate := filepath.Clean(filepath.Join(dir, name))
		if _, err := os.Stat(candidate); err == nil {
			return candidate, nil
		}
	}
	return "", fmt.Errorf("%s file not found", name)
}

func loadDotEnv(path string) error {
	file, err := os.Open(path)
	if err != nil {
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if strings.HasPrefix(line, "export ") {
			line = strings.TrimSpace(strings.TrimPrefix(line, "export "))
		}
		parts := strings.SplitN(line, "=", 2)
		if len(parts) != 2 {
			continue
		}
		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])
		value = strings.Trim(value, `"'`)
		if key != "" {
			_ = os.Setenv(key, value)
		}
	}
	return scanner.Err()
}

func collectExpiredUsers(apiKey, baseURL string, cutoff time.Time, maxProcessed int) ([]expiredUserRow, int, error) {
	results := make([]expiredUserRow, 0)
	count := 0
	offset := 0
	client := newHTTPClient()

	if maxProcessed > 0 {
		fmt.Printf("Quick-scan limiter active: will stop after %d users\n", maxProcessed)
	}

	for {
		fmt.Printf("Fetching users: offset=%d limit=%d\n", offset, pageLimit)
		data, err := fetchUsersPage(client, apiKey, baseURL, offset, pageLimit)
		if err != nil {
			return results, count, err
		}

		fmt.Printf("  Received %d users\n", len(data.Users))
		if len(data.Users) == 0 {
			break
		}

		for _, user := range data.Users {
			count++
			if debugSample && count == 1 {
				b, _ := json.MarshalIndent(user, "", "  ")
				fmt.Println("\n--- Debug: sample user (first returned) ---")
				fmt.Println(string(b))
			}
			if maxProcessed > 0 && count > maxProcessed {
				fmt.Printf("Reached scanning limit (%d) - stopping early\n", maxProcessed)
				return results, count - 1, nil
			}

			primaryID := firstString(user, "primary_id", "primaryId", "primary")
			firstName := firstString(user, "first_name", "firstName")
			lastName := firstString(user, "last_name", "lastName")
			expiration := extractUserExpiration(user)
			email := extractEmailFromContact(user["contact_info"])
			patronType := extractPatronType(user["user_group"])

			if expiration != nil && !expiration.After(cutoff) {
				results = append(results, expiredUserRow{
					PrimaryID:      primaryID,
					FirstName:      firstName,
					LastName:       lastName,
					Email:          email,
					PatronType:     patronType,
					ExpirationDate: expiration.Format(time.RFC3339),
				})
			}
		}

		offset += pageLimit
		if data.TotalRecordCount > 0 && offset >= data.TotalRecordCount {
			break
		}
		if len(data.Users) < pageLimit {
			break
		}

		time.Sleep(200 * time.Millisecond)
	}

	return results, count, nil
}

func newHTTPClient() *http.Client {
	transport := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	return &http.Client{
		Timeout:   requestTimeout,
		Transport: transport,
	}
}

func fetchUsersPage(client *http.Client, apiKey, baseURL string, offset, limit int) (*usersResponse, error) {
	endpoint, err := url.Parse(strings.TrimRight(baseURL, "/") + "/almaws/v1/users")
	if err != nil {
		return nil, err
	}

	q := endpoint.Query()
	q.Set("limit", strconv.Itoa(limit))
	q.Set("offset", strconv.Itoa(offset))
	q.Set("format", "json")
	q.Set("expand", "full")
	endpoint.RawQuery = q.Encode()

	req, err := http.NewRequest(http.MethodGet, endpoint.String(), nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "apikey "+apiKey)
	req.Header.Set("Accept", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch users: %d %s", resp.StatusCode, strings.TrimSpace(string(body)))
	}

	var data usersResponse
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	return &data, nil
}

func parseAlmaDate(s string) *time.Time {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil
	}

	if len(s) == len("2006-01-02Z") && strings.HasSuffix(s, "Z") {
		s = strings.TrimSuffix(s, "Z")
	}

	layouts := []string{
		"2006-01-02",
		"2006-01-02Z07:00",
		time.RFC3339Nano,
		time.RFC3339,
		"2006-01-02T15:04:05",
		"2006-01-02T15:04:05.000",
	}

	for _, layout := range layouts {
		if t, err := time.Parse(layout, s); err == nil {
			return &t
		}
	}

	if t, err := time.Parse("2006-01-02T15:04:05Z07:00", s); err == nil {
		return &t
	}

	return nil
}

func extractUserExpiration(user map[string]any) *time.Time {
	if expiration := parseAlmaDate(firstString(user, "expiration_date", "expiry_date", "expirationDate", "expiryDate", "expiration")); expiration != nil {
		return expiration
	}

	roles, ok := user["user_role"].([]any)
	if !ok {
		return nil
	}

	var latest *time.Time
	for _, role := range roles {
		roleMap, ok := role.(map[string]any)
		if !ok {
			continue
		}

		expiration := parseAlmaDate(firstString(roleMap, "expiration_date", "expiry_date", "expirationDate", "expiryDate", "expiration"))
		if expiration == nil {
			continue
		}
		if latest == nil || expiration.After(*latest) {
			latest = expiration
		}
	}

	return latest
}

func extractPatronType(userGroup any) string {
	m, ok := userGroup.(map[string]any)
	if !ok {
		return ""
	}

	if value := stringValue(m["value"]); value != "" {
		return value
	}
	return stringValue(m["desc"])
}

func extractEmailFromContact(contactInfo any) string {
	m, ok := contactInfo.(map[string]any)
	if !ok {
		return ""
	}

	rawEmails, ok := m["email"]
	if !ok {
		return ""
	}

	emails, ok := rawEmails.([]any)
	if !ok || len(emails) == 0 {
		return ""
	}

	for _, item := range emails {
		emailMap, ok := item.(map[string]any)
		if !ok {
			continue
		}
		if preferred, ok := emailMap["preferred"].(bool); ok && preferred {
			return stringValue(emailMap["email_address"])
		}
	}

	if emailMap, ok := emails[0].(map[string]any); ok {
		return stringValue(emailMap["email_address"])
	}
	return ""
}

func firstString(m map[string]any, keys ...string) string {
	for _, key := range keys {
		if val, ok := m[key]; ok {
			if s := stringValue(val); s != "" {
				return s
			}
		}
	}
	return ""
}

func stringValue(v any) string {
	switch t := v.(type) {
	case string:
		return t
	case fmt.Stringer:
		return t.String()
	case nil:
		return ""
	default:
		return fmt.Sprintf("%v", t)
	}
}

func ensureOutputDir() (string, error) {
	cwd, err := os.Getwd()
	if err != nil {
		return "", err
	}

	candidates := []string{
		filepath.Join(cwd, "../outputs"),
		filepath.Join(cwd, "outputs"),
		filepath.Join(cwd, "patron-checker/outputs"),
	}

	for _, candidate := range candidates {
		cleaned := filepath.Clean(candidate)
		if _, err := os.Stat(filepath.Dir(cleaned)); err == nil {
			if err := os.MkdirAll(cleaned, 0o755); err == nil {
				if rel, err := filepath.Rel(cwd, cleaned); err == nil {
					return rel, nil
				}
				return cleaned, nil
			}
		}
	}

	return "", fmt.Errorf("unable to create outputs directory")
}

func saveCSV(rows []expiredUserRow, environment string) (string, error) {
	outputDir, err := ensureOutputDir()
	if err != nil {
		return "", err
	}

	filename := filepath.Join(outputDir, fmt.Sprintf("expired_users_%s_%s.csv", environment, time.Now().Format("20060102_150405")))
	file, err := os.Create(filename)
	if err != nil {
		return "", err
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	if err := writer.Write([]string{"primary_id", "first_name", "last_name", "email", "patron_type", "expiration_date"}); err != nil {
		return "", err
	}
	for _, row := range rows {
		if err := writer.Write([]string{row.PrimaryID, row.FirstName, row.LastName, row.Email, row.PatronType, row.ExpirationDate}); err != nil {
			return "", err
		}
	}

	return filename, writer.Error()
}

func saveJSON(rows []expiredUserRow, environment string) (string, error) {
	outputDir, err := ensureOutputDir()
	if err != nil {
		return "", err
	}

	filename := filepath.Join(outputDir, fmt.Sprintf("expired_users_%s_%s.json", environment, time.Now().Format("20060102_150405")))
	file, err := os.Create(filename)
	if err != nil {
		return "", err
	}
	defer file.Close()

	encoder := json.NewEncoder(file)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(rows); err != nil {
		return "", err
	}
	return filename, nil
}

